File: backend/app/api/forum.py
from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy import select, desc
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy.orm import selectinload
from app.database import get_db
from app.models.forum import ForumPost, ForumComment, Notification
from app.models.user import User
from app.api.auth import get_current_user
from pydantic import BaseModel
from datetime import datetime

router = APIRouter()

class PostIn(BaseModel):
    title: str
    content: str

class PostOut(BaseModel):
    id: int
    user_id: int
    username: str
    title: str
    content: str
    created_at: datetime

    class Config:
        from_attributes = True

@router.get("/posts")
async def list_posts(
    page: int = Query(1, ge=1),
    limit: int = Query(20, ge=1, le=100),
    db: AsyncSession = Depends(get_db),
):
    offset = (page - 1) * limit
    result = await db.execute(
        select(ForumPost)
        .options(selectinload(ForumPost.user))
        .order_by(desc(ForumPost.created_at))
        .offset(offset)
        .limit(limit)
    )
    posts = result.scalars().all()
    count_result = await db.execute(select(ForumPost))
    total = len(count_result.scalars().all())
    # Transform posts to include username
    items = []
    for p in posts:
        items.append({
            "id": p.id,
            "user_id": p.user_id,
            "username": p.user.username,
            "title": p.title,
            "content": p.content,
            "created_at": p.created_at,
        })
    return {
        "items": items,
        "total": total,
        "page": page,
        "limit": limit
    }

@router.post("/posts", response_model=PostOut)
async def create_post(
    data: PostIn,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    if current_user.is_muted:
        raise HTTPException(status_code=403, detail="您已被禁言，无法发帖")
    post = ForumPost(
        user_id=current_user.id,
        title=data.title,
        content=data.content,
    )
    db.add(post)
    await db.commit()
    await db.refresh(post)
    return {
        "id": post.id,
        "user_id": post.user_id,
        "username": current_user.username,
        "title": post.title,
        "content": post.content,
        "created_at": post.created_at,
    }

@router.delete("/posts/{post_id}")
async def delete_post(
    post_id: int,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    result = await db.execute(select(ForumPost).where(ForumPost.id == post_id))
    post = result.scalar_one_or_none()
    if not post:
        raise HTTPException(status_code=404, detail="帖子不存在")
    if post.user_id != current_user.id and not current_user.is_admin:
        raise HTTPException(status_code=403, detail="无权删除此帖子")
    await db.delete(post)
    await db.commit()
    return {"message": "删除成功"}

class CommentIn(BaseModel):
    content: str

class CommentOut(BaseModel):
    id: int
    post_id: int
    user_id: int
    username: str
    content: str
    created_at: datetime

@router.get("/posts/{post_id}/comments")
async def list_comments(post_id: int, db: AsyncSession = Depends(get_db)):
    result = await db.execute(
        select(ForumComment)
        .options(selectinload(ForumComment.user))
        .where(ForumComment.post_id == post_id)
        .order_by(ForumComment.created_at)
    )
    comments = result.scalars().all()
    return [{
        "id": c.id,
        "post_id": c.post_id,
        "user_id": c.user_id,
        "username": c.user.username,
        "content": c.content,
        "created_at": c.created_at,
    } for c in comments]

@router.post("/posts/{post_id}/comments", response_model=CommentOut)
async def create_comment(
    post_id: int,
    data: CommentIn,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    if current_user.is_muted:
        raise HTTPException(status_code=403, detail="您已被禁言，无法评论")
    # 确认帖子存在
    post_result = await db.execute(select(ForumPost).where(ForumPost.id == post_id))
    post = post_result.scalar_one_or_none()
    if not post:
        raise HTTPException(status_code=404, detail="帖子不存在")
    comment = ForumComment(
        post_id=post_id,
        user_id=current_user.id,
        content=data.content,
    )
    db.add(comment)
    await db.commit()
    await db.refresh(comment)
    # 通知帖子作者（如果评论者不是作者本人）
    if post.user_id != current_user.id:
        notif = Notification(
            user_id=post.user_id,
            type="comment_reply",
            title="有人回复了你的帖子",
            content=f"{current_user.username} 回复了「{post.title}」",
        )
        db.add(notif)
        await db.commit()
    return {
        "id": comment.id,
        "post_id": comment.post_id,
        "user_id": comment.user_id,
        "username": current_user.username,
        "content": comment.content,
        "created_at": comment.created_at,
    }

@router.delete("/comments/{comment_id}")
async def delete_comment(
    comment_id: int,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    result = await db.execute(select(ForumComment).where(ForumComment.id == comment_id))
    comment = result.scalar_one_or_none()
    if not comment:
        raise HTTPException(status_code=404, detail="评论不存在")
    if comment.user_id != current_user.id and not current_user.is_admin:
        raise HTTPException(status_code=403, detail="无权删除此评论")
    await db.delete(comment)
    await db.commit()
    return {"message": "删除成功"}

class NotificationIn(BaseModel):
    title: str
    content: str

class NotificationOut(BaseModel):
    id: int
    user_id: int
    type: str
    title: str
    content: str
    is_read: bool
    created_at: datetime

@router.get("/notifications")
async def get_notifications(db: AsyncSession = Depends(get_db), current_user: User = Depends(get_current_user)):
    result = await db.execute(
        select(Notification)
        .where(Notification.user_id == current_user.id)
        .order_by(desc(Notification.created_at))
        .limit(50)
    )
    notifications = result.scalars().all()
    return [{"id": n.id, "user_id": n.user_id, "type": n.type, "title": n.title, "content": n.content, "is_read": n.is_read, "created_at": n.created_at} for n in notifications]

@router.put("/notifications/{notif_id}/read")
async def mark_read(notif_id: int, db: AsyncSession = Depends(get_db), current_user: User = Depends(get_current_user)):
    result = await db.execute(select(Notification).where(Notification.id == notif_id, Notification.user_id == current_user.id))
    n = result.scalar_one_or_none()
    if not n:
        raise HTTPException(status_code=404, detail="通知不存在")
    n.is_read = True
    await db.commit()
    return {"message": "已读"}

@router.put("/notifications/read-all")
async def mark_all_read(db: AsyncSession = Depends(get_db), current_user: User = Depends(get_current_user)):
    result = await db.execute(
        select(Notification).where(Notification.user_id == current_user.id, Notification.is_read == False)
    )
    unread = result.scalars().all()
    for n in unread:
        n.is_read = True
    await db.commit()
    return {"message": f"已标记 {len(unread)} 条为已读"}

# 管理员发布公告
@router.post("/admin/announcements", response_model=NotificationOut)
async def create_announcement(
    data: NotificationIn,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    if not current_user.is_admin:
        raise HTTPException(status_code=403, detail="需要管理员权限")
    result = await db.execute(select(User.id))
    all_user_ids = [r[0] for r in result.fetchall()]
    notifications = []
    for uid in all_user_ids:
        n = Notification(
            user_id=uid,
            type="announcement",
            title=data.title,
            content=data.content,
        )
        db.add(n)
        notifications.append(n)
    await db.commit()
    for n in notifications:
        await db.refresh(n)
    return {
        "id": notifications[0].id,
        "user_id": notifications[0].user_id,
        "type": notifications[0].type,
        "title": notifications[0].title,
        "content": notifications[0].content,
        "is_read": notifications[0].is_read,
        "created_at": notifications[0].created_at,
    }

File: internal/handler/forum.go
package handler

import (
	"database/sql"
	"net/http"
	"strconv"
	"time"

	"github.com/labstack/echo/v4"
)

type ForumHandler struct {
	DB *sql.DB
}

// ── Schemas ──────────────────────────────────────────────────────────────────

type PostIn struct {
	Title   string `json:"title"`
	Content string `json:"content"`
}

type PostOut struct {
	ID        int64     `json:"id"`
	UserID    int64     `json:"user_id"`
	Username  string    `json:"username"`
	Title     string    `json:"title"`
	Content   string    `json:"content"`
	CreatedAt time.Time `json:"created_at"`
}

type CommentIn struct {
	Content string `json:"content"`
}

type CommentOut struct {
	ID        int64     `json:"id"`
	PostID    int64     `json:"post_id"`
	UserID    int64     `json:"user_id"`
	Username  string    `json:"username"`
	Content   string    `json:"content"`
	CreatedAt time.Time `json:"created_at"`
}

type PostListOut struct {
	Items []PostOut `json:"items"`
	Total int       `json:"total"`
	Page  int       `json:"page"`
	Limit int       `json:"limit"`
}

type NotifOut struct {
	ID        int64     `json:"id"`
	UserID    int64     `json:"user_id"`
	Type      string    `json:"type"`
	Title     string    `json:"title"`
	Content   string    `json:"content,omitempty"`
	IsRead    bool      `json:"is_read"`
	CreatedAt time.Time `json:"created_at"`
}

// ── GET /api/forum/posts ─────────────────────────────────────────────────────

func (h *ForumHandler) ListPosts(c echo.Context) error {
	page, _ := strconv.Atoi(c.QueryParam("page"))
	if page < 1 {
		page = 1
	}
	limit, _ := strconv.Atoi(c.QueryParam("limit"))
	if limit < 1 || limit > 100 {
		limit = 20
	}
	offset := (page - 1) * limit

	var total int
	h.DB.QueryRow(`SELECT COUNT(*) FROM forum_posts`).Scan(&total)

	rows, err := h.DB.Query(`
		SELECT p.id, p.user_id, p.title, p.content, p.created_at, u.username
		FROM forum_posts p
		LEFT JOIN users u ON p.user_id = u.id
		ORDER BY p.created_at DESC
		LIMIT $1 OFFSET $2`, limit, offset)
	if err != nil {
		return echo.NewHTTPError(http.StatusInternalServerError, err.Error())
	}
	defer rows.Close()

	items := []PostOut{}
	for rows.Next() {
		var p PostOut
		var createdAt time.Time
		if err := rows.Scan(&p.ID, &p.UserID, &p.Title, &p.Content, &createdAt, &p.Username); err != nil {
			continue
		}
		p.CreatedAt = createdAt
		items = append(items, p)
	}

	return c.JSON(http.StatusOK, PostListOut{Items: items, Total: total, Page: page, Limit: limit})
}

// ── POST /api/forum/posts ───────────────────────────────────────────────────

func (h *ForumHandler) CreatePost(c echo.Context) error {
	userID := c.Get("user_id").(int64)

	var isMuted bool
	h.DB.QueryRow(`SELECT is_muted FROM users WHERE id = $1`, userID).Scan(&isMuted)
	if isMuted {
		return echo.NewHTTPError(http.StatusForbidden, "您已被禁言，无法发帖")
	}

	var req PostIn
	if err := c.Bind(&req); err != nil {
		return echo.NewHTTPError(http.StatusBadRequest, "invalid request")
	}
	if req.Title == "" || req.Content == "" {
		return echo.NewHTTPError(http.StatusBadRequest, "title and content are required")
	}

	var p PostOut
	var createdAt time.Time
	err := h.DB.QueryRow(`
		INSERT INTO forum_posts (user_id, title, content)
		VALUES ($1, $2, $3)
		RETURNING id, user_id, title, content, created_at`,
		userID, req.Title, req.Content,
	).Scan(&p.ID, &p.UserID, &p.Title, &p.Content, &createdAt)
	if err != nil {
		return echo.NewHTTPError(http.StatusInternalServerError, err.Error())
	}
	p.CreatedAt = createdAt

	var username string
	h.DB.QueryRow(`SELECT username FROM users WHERE id = $1`, userID).Scan(&username)
	p.Username = username

	return c.JSON(http.StatusCreated, p)
}

// ── DELETE /api/forum/posts/:id ────────────────────────────────────────────

func (h *ForumHandler) DeletePost(c echo.Context) error {
	userID := c.Get("user_id").(int64)
	postID, err := strconv.ParseInt(c.Param("id"), 10, 64)
	if err != nil {
		return echo.NewHTTPError(http.StatusBadRequest, "invalid post id")
	}

	var isAdmin bool
	h.DB.QueryRow(`SELECT is_admin FROM users WHERE id = $1`, userID).Scan(&isAdmin)

	var ownerID int64
	h.DB.QueryRow(`SELECT user_id FROM forum_posts WHERE id = $1`, postID).Scan(&ownerID)
	if ownerID == 0 {
		return echo.NewHTTPError(http.StatusNotFound, "帖子不存在")
	}
	if ownerID != userID && !isAdmin {
		return echo.NewHTTPError(http.StatusForbidden, "无权删除此帖子")
	}

	h.DB.Exec(`DELETE FROM forum_posts WHERE id = $1`, postID)
	return c.JSON(http.StatusOK, map[string]string{"message": "删除成功"})
}

// ── GET /api/forum/posts/:id/comments ─────────────────────────────────────

func (h *ForumHandler) ListComments(c echo.Context) error {
	postID, err := strconv.ParseInt(c.Param("id"), 10, 64)
	if err != nil {
		return echo.NewHTTPError(http.StatusBadRequest, "invalid post id")
	}

	rows, err := h.DB.Query(`
		SELECT c.id, c.post_id, c.user_id, c.content, c.created_at, u.username
		FROM forum_comments c
		LEFT JOIN users u ON c.user_id = u.id
		WHERE c.post_id = $1
		ORDER BY c.created_at ASC`, postID)
	if err != nil {
		return echo.NewHTTPError(http.StatusInternalServerError, err.Error())
	}
	defer rows.Close()

	items := []CommentOut{}
	for rows.Next() {
		var c CommentOut
		var createdAt time.Time
		if err := rows.Scan(&c.ID, &c.PostID, &c.UserID, &c.Content, &createdAt, &c.Username); err != nil {
			continue
		}
		c.CreatedAt = createdAt
		items = append(items, c)
	}
	return c.JSON(http.StatusOK, items)
}

// ── POST /api/forum/posts/:id/comments ─────────────────────────────────────

func (h *ForumHandler) CreateComment(c echo.Context) error {
	userID := c.Get("user_id").(int64)
	postID, err := strconv.ParseInt(c.Param("id"), 10, 64)
	if err != nil {
		return echo.NewHTTPError(http.StatusBadRequest, "invalid post id")
	}

	var isMuted bool
	h.DB.QueryRow(`SELECT is_muted FROM users WHERE id = $1`, userID).Scan(&isMuted)
	if isMuted {
		return echo.NewHTTPError(http.StatusForbidden, "您已被禁言，无法评论")
	}

	var postOwnerID int64
	var postTitle string
	err = h.DB.QueryRow(`SELECT user_id, title FROM forum_posts WHERE id = $1`, postID).Scan(&postOwnerID, &postTitle)
	if err == sql.ErrNoRows {
		return echo.NewHTTPError(http.StatusNotFound, "帖子不存在")
	}

	var req CommentIn
	if err := c.Bind(&req); err != nil {
		return echo.NewHTTPError(http.StatusBadRequest, "invalid request")
	}
	if req.Content == "" {
		return echo.NewHTTPError(http.StatusBadRequest, "content is required")
	}

	var commentID, commentUserID int64
	var commentContent string
	var commentCreatedAt time.Time
	err = h.DB.QueryRow(`
		INSERT INTO forum_comments (post_id, user_id, content)
		VALUES ($1, $2, $3)
		RETURNING id, user_id, content, created_at`,
		postID, userID, req.Content,
	).Scan(&commentID, &commentUserID, &commentContent, &commentCreatedAt)
	if err != nil {
		return echo.NewHTTPError(http.StatusInternalServerError, err.Error())
	}

	var username string
	h.DB.QueryRow(`SELECT username FROM users WHERE id = $1`, userID).Scan(&username)

	// Notify post author
	if postOwnerID != userID {
		h.DB.Exec(`
			INSERT INTO notifications (user_id, type, title, content)
			VALUES ($1, 'comment_reply', '有人回复了你的帖子', $2)`,
			postOwnerID, username+" 回复了「"+postTitle+"」")
	}

	return c.JSON(http.StatusCreated, CommentOut{
		ID:        commentID,
		PostID:    postID,
		UserID:    commentUserID,
		Username:  username,
		Content:   commentContent,
		CreatedAt: commentCreatedAt,
	})
}

// ── DELETE /api/forum/comments/:id ─────────────────────────────────────────

func (h *ForumHandler) DeleteComment(c echo.Context) error {
	userID := c.Get("user_id").(int64)
	commentID, err := strconv.ParseInt(c.Param("id"), 10, 64)
	if err != nil {
		return echo.NewHTTPError(http.StatusBadRequest, "invalid comment id")
	}

	var isAdmin bool
	h.DB.QueryRow(`SELECT is_admin FROM users WHERE id = $1`, userID).Scan(&isAdmin)

	var commentOwnerID int64
	h.DB.QueryRow(`SELECT user_id FROM forum_comments WHERE id = $1`, commentID).Scan(&commentOwnerID)
	if commentOwnerID == 0 {
		return echo.NewHTTPError(http.StatusNotFound, "评论不存在")
	}
	if commentOwnerID != userID && !isAdmin {
		return echo.NewHTTPError(http.StatusForbidden, "无权删除此评论")
	}

	h.DB.Exec(`DELETE FROM forum_comments WHERE id = $1`, commentID)
	return c.JSON(http.StatusOK, map[string]string{"message": "删除成功"})
}

// ── GET /api/forum/notifications ───────────────────────────────────────────

func (h *ForumHandler) GetNotifications(c echo.Context) error {
	userID := c.Get("user_id").(int64)

	rows, err := h.DB.Query(`
		SELECT id, user_id, type, title, content, is_read, created_at
		FROM notifications
		WHERE user_id = $1
		ORDER BY created_at DESC
		LIMIT 50`, userID)
	if err != nil {
		return echo.NewHTTPError(http.StatusInternalServerError, err.Error())
	}
	defer rows.Close()

	items := []NotifOut{}
	for rows.Next() {
		var n NotifOut
		var content sql.NullString
		var createdAt time.Time
		if err := rows.Scan(&n.ID, &n.UserID, &n.Type, &n.Title, &content, &n.IsRead, &createdAt); err != nil {
			continue
		}
		if content.Valid {
			n.Content = content.String
		}
		n.CreatedAt = createdAt
		items = append(items, n)
	}
	return c.JSON(http.StatusOK, items)
}

// ── PUT /api/forum/notifications/:id/read ──────────────────────────────────

func (h *ForumHandler) MarkRead(c echo.Context) error {
	userID := c.Get("user_id").(int64)
	notifID, err := strconv.ParseInt(c.Param("id"), 10, 64)
	if err != nil {
		return echo.NewHTTPError(http.StatusBadRequest, "invalid notification id")
	}

	res, err := h.DB.Exec(`
		UPDATE notifications SET is_read = TRUE
		WHERE id = $1 AND user_id = $2`, notifID, userID)
	if err != nil {
		return echo.NewHTTPError(http.StatusInternalServerError, err.Error())
	}
	rows, _ := res.RowsAffected()
	if rows == 0 {
		return echo.NewHTTPError(http.StatusNotFound, "通知不存在")
	}
	return c.JSON(http.StatusOK, map[string]string{"message": "已读"})
}

// ── PUT /api/forum/notifications/read-all ─────────────────────────────────

func (h *ForumHandler) MarkAllRead(c echo.Context) error {
	userID := c.Get("user_id").(int64)

	res, err := h.DB.Exec(`
		UPDATE notifications SET is_read = TRUE
		WHERE user_id = $1 AND is_read = FALSE`, userID)
	if err != nil {
		return echo.NewHTTPError(http.StatusInternalServerError, err.Error())
	}
	rows, _ := res.RowsAffected()
	return c.JSON(http.StatusOK, map[string]string{
		"message": "已标记 " + strconv.FormatInt(rows, 10) + " 条为已读",
	})
}

// ── POST /api/forum/admin/announcements ────────────────────────────────────

func (h *ForumHandler) CreateAnnouncement(c echo.Context) error {
	userID := c.Get("user_id").(int64)

	var isAdmin bool
	h.DB.QueryRow(`SELECT is_admin FROM users WHERE id = $1`, userID).Scan(&isAdmin)
	if !isAdmin {
		return echo.NewHTTPError(http.StatusForbidden, "需要管理员权限")
	}

	var req struct {
		Title   string `json:"title"`
		Content string `json:"content"`
	}
	if err := c.Bind(&req); err != nil {
		return echo.NewHTTPError(http.StatusBadRequest, "invalid request")
	}
	if req.Title == "" {
		return echo.NewHTTPError(http.StatusBadRequest, "title is required")
	}

	rows, err := h.DB.Query(`SELECT id FROM users`)
	if err != nil {
		return echo.NewHTTPError(http.StatusInternalServerError, err.Error())
	}
	var firstID int64
	for rows.Next() {
		var uid int64
		rows.Scan(&uid)
		if _, err := h.DB.Exec(`
			INSERT INTO notifications (user_id, type, title, content)
			VALUES ($1, 'announcement', $2, $3)`, uid, req.Title, req.Content); err != nil {
			return echo.NewHTTPError(http.StatusInternalServerError, "failed to create notification: "+err.Error())
		}
		if firstID == 0 {
			firstID = uid
		}
	}
	rows.Close()

	return c.JSON(http.StatusOK, map[string]interface{}{
		"id":       0,
		"user_id":  firstID,
		"type":     "announcement",
		"title":    req.Title,
		"content":  req.Content,
		"is_read": false,
	})
}
